src/ion_data/mod.rs

```diff
@@ -0,0 +1,76 @@
+mod ion_eq;
+mod ion_ord;
+
+use std::cmp::Ordering;
+use std::fmt::{Display, Formatter};
+
+pub(crate) use ion_eq::IonEq;
+pub(crate) use ion_ord::IonOrd;
+
+/// A wrapper for lifting Ion compatible data into using Ion oriented comparisons (versus the Rust
+/// value semantics). This enables the default semantics to be what a Rust user expects for native
+/// values, but allows a user to opt-in to Ion's structural equivalence/order.
+///
+/// Anything that implements [`IonEq`] can be converted into [`IonData`]. [`IonData`] itself does not
+/// implement [`IonEq`] so that it is impossible to create, for example, `IonData<IonData<Element>>`.
+///
+/// [`Hash`] and [`Ord`] are not guaranteed to be implemented for all [`IonData`], but when they are,
+/// they are required to be consistent with [`IonEq`] (and [`Eq`]).
+///
+/// __WARNING__—The Ion specification does _not_ define a total ordering over all Ion values. Do
+/// not depend on getting any particular result from [Ord]. Use it only as an opaque total ordering
+/// over all [IonData]. _The algorithm used for [Ord] may change between versions._
+#[derive(Debug, Clone)]
+pub struct IonData<T: IonEq>(T);
+
+impl<T: IonEq> IonData<T> {
+    // Checks if two values are equal according to Ion's structural equivalence.
+    pub fn eq(a: &T, b: &T) -> bool {
+        T::ion_eq(a, b)
+    }
```

Comment:
It just occurred to me that something like this might be a bit more flexible (works on anything that `AsRef<T>`)

```suggestion
    /// Checks if two values are equal according to Ion's structural equivalence.
    pub fn eq<R>(a: R, b: R) -> bool where R: AsRef<T> {
        T::ion_eq(a.as_ref(), b.as_ref())
    }
```

Reply:
Any thoughts about using this vs. putting `AsRef` into the function signature?
```rust
impl <T: IonEq, R: Deref<Target = T>> IonEq for R {
    fn ion_eq(&self, other: &Self) -> bool {
        // ...
    }
}
```

Using `AsRef` in the function signature means that in some cases, we need to start adding type annotations (turbofish, not ion) when there's more than one `AsRef` implementation for a given type.

Do you think that implementing `IonEq` for smart pointers would capture your intent, though in a different way?

Reply:
Does this work for the case where we don't want to expose `IonEq`, would external users have to use `Into<IonData>` to get at this?

Reply:
> would external users have to use `Into<IonData>` to get at this?

No, they would not. As long as something implements `IonEq`, external users can (without knowing about `IonEq`) use `IonData::eq(foo, bar)` if all they want to do is check for Ion equivalence.

Reply:
I think I would be good with this formulation--the net effect is basically the same I think.

Reply:
Oh shoot. I was wrong. I still had `IonEq` as `pub` instead of `pub(crate)`. I think `IonEq` might need to be `pub`, but we can prevent other implementors by using the [sealed traits](https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed) pattern. I'm going to play around with it a little more to see if I can figure out something that works nicely.